### PR TITLE
[WIP] Refactor cache section in tempodb/backend

### LINKED ISSUE #373: Refactor cache section in tempodb/backend to use Cache interface
The Cache section of `tempodb/backend` has a lot of copied code between `disk_cache`, `memcached` & `redis`. 

This is because each of the cache backends are implementing the entire `backend.Reader` & `backend.Writer` interfaces. Instead, we should split `cache` into a separate package (that lives in `tempodb/backend/cache`) and should accept a cache implementation that implements a smaller interface that interacts with the respective datastores. A rough draft looks something like this

```
type Cache interface {
    func Get(ctx context.Context, key string) []byte
    func Set(ctx context.Context, key string, val []byte)
}
```

Each of the caches should live within `tempodb/backend/cache/disk_cache`, `tempodb/backend/cache/memcached` and `tempodb/backend/cache/redis` respectively and just implement this small interface described above.

This will reduce repetition and keep all of the testing in one place.

Came up in https://github.com/grafana/tempo/pull/354#discussion_r529283403

### LINKED ISSUE #424: Delete Disk Cache
The disk cache only mostly worked when we used it, but we have since added support for memcached and redis.

Let's remove this as we don't want to support it and it had issues at large scale anyway.

### PR DESCRIPTION
**Pull request by dgzlopes**
**What this PR does**:
TBD

**Which issue(s) this PR fixes**:
Fixes #373
Fixes #424 

**Checklist**
- [ ] Tests updated ---> Not yet!
- [ ] Documentation added ---> Not yet!
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

cc @annanay25

### CODE REVIEW

**Comment by annanay25** [comment]
Great work @dgzlopes! 🎉  I'll take a look tomorrow, but this is definitely in the right direction!

File: tempodb/backend/cache/memcached/memcached.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
+package memcached
```
Comment (annanay25) [inline tempodb/backend/cache/memcached/memcached.go:1]: We can add a `memcached_test.go` with just a single line:

```
var _ (cache.Cache) = (*Cache)(nil)
```

similarly for redis.

File: tempodb/backend/cache/memcached/memcached.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/grafana/tempo/tempodb/backend/cache"
+)
+
+type Config struct {
+	ClientConfig cortex_cache.MemcachedClientConfig `yaml:",inline"`
+
+	TTL time.Duration `yaml:"ttl"`
+}
+
+type Cache struct {
```
Comment (annanay25) [inline tempodb/backend/cache/memcached/memcached.go:20]: Let's name this `Client`. That way when we create the client the signature will be `memcache.NewClient()`., which also helps reduce confusion. Similarly for Redis.

**Review by annanay25 — COMMENTED** [review]
Feeling really good about this PR, great work @dgzlopes :)

Let's add the tests and we can land this soon!

**Comment by dgzlopes** [comment]
Closing in favor of #485

**Pull request closed without merging**